crates/sui-core/src/event_handler.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::{ArcWrapper, AuthorityStore};
+use crate::streamer::Streamer;
+use futures::future::join_all;
+use move_bytecode_utils::module_cache::SyncModuleCache;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use sui_types::{
+    error::{SuiError, SuiResult},
+    event::{Event, EventEnvelope},
+    messages::TransactionEffects,
+};
+use tokio::sync::mpsc::{self, Sender};
+
+use tracing::{debug, error};
+
+const EVENT_DISPATCH_BUFFER_SIZE: usize = 1000;
+
+pub fn get_unixtime_ms() -> u128 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Travelling in time machine")
+        .as_millis()
+}
+
+pub struct EventHandler {
+    module_cache: SyncModuleCache<ArcWrapper<AuthorityStore>>,
+    streamer_queue: Sender<EventEnvelope>,
+}
+
+impl EventHandler {
+    pub fn new(validator_store: Arc<AuthorityStore>) -> Self {
+        let (tx, rx) = mpsc::channel::<EventEnvelope>(EVENT_DISPATCH_BUFFER_SIZE);
+        Streamer::spawn(rx);
+        Self {
+            module_cache: SyncModuleCache::new(ArcWrapper(validator_store)),
+            streamer_queue: tx,
+        }
+    }
+
+    pub async fn process_events(&self, effects: &TransactionEffects) {
+        let futures = effects.events.iter().map(|e| self.process_event(e));
+        let results = join_all(futures).await;
+        for r in &results {
+            if let Err(e) = r {
+                error!(error =? e, "Failed to send EventEnvolope to dispatch");
+            }
+        }
+    }
+
+    pub async fn process_event(&self, event: &Event) -> SuiResult {
+        let envolope = match event {
+            Event::MoveEvent {
+                type_: _,
+                contents: _,
+            } => {
+                debug!(event =? event, "Process MoveEvent.");
+                match event.extract_move_struct(&self.module_cache) {
```

Comment:
So, getting the move struct out of the event will also be needed by the path to write to event store also.  Could we do the event handler in a reactive / async stream style, then we can just have a chain of event processing steps in the stream?  The first one could be adding the move struct, or we could add it via populating a OnceCell based field in MoveEvent for the struct.  I could then use the output of this first step in the EventStore.

Reply:
I see, maybe I can add `MoveStruct` in `EventEnvolope` too if it's useful afterwards

regarding streaming, I didn't make it that way because the pre-processing part (mainly`extract move struct`) seems to be trivial. Do you just prefer the streaming style or do you see other advantages by streaming earlier on? (In case you missed it, I was thinking to interact with event store in line 73)

The event dissemination is designed in a streaming way, see line 81-86

Reply:
So basically the streaming style - well I guess part of it is preference, but part of it is a easy way to bulk ingest events and take advantage of processing multiple events.  Like, we don't want to just ingest one event at a time, but at minimum to ingest all the events for a single TX.  Better yet though is to ingest many events at the same time.

Also if you handle each type of event differently how can you be sure that the order is preserved in the stream?   Keeping things as a stream makes sure we still preserve ordering.